infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPods.java

```diff
@@ -554,22 +554,36 @@ public void delete() throws InfrastructureException {
   }
 
   private CompletableFuture<Void> doDelete(String name) throws InfrastructureException {
+    Watch toCloseOnException = null;
     try {
       final PodResource<Pod, DoneablePod> podResource =
           clientFactory.create(workspaceId).pods().inNamespace(namespace).withName(name);
-      final CompletableFuture<Void> deleteFuture = new CompletableFuture<>();
+      CompletableFuture<Void> deleteFuture = new CompletableFuture<>();
       final Watch watch = podResource.watch(new DeleteWatcher(deleteFuture));
-
-      podResource.delete();
-      return deleteFuture.whenComplete(
-          (v, e) -> {
-            if (e != null) {
-              LOG.warn("Failed to remove pod {} cause {}", name, e.getMessage());
-            }
-            watch.close();
-          });
+      toCloseOnException = watch;
+      deleteFuture =
+          deleteFuture.whenComplete(
+              (v, e) -> {
+                if (e != null) {
+                  LOG.warn("Failed to remove pod {} cause {}", name, e.getMessage());
+                }
+                watch.close();
+              });
+      Boolean deleteSucceeded = podResource.delete();
+      if (deleteSucceeded == null || !deleteSucceeded) {
+        watch.close();
```

Comment:
I think, in a case when a pod doesn't exist, delete future will be completed exceptionally with message `Webscoket connection is closed. But event about removing is not received.`. And I would say that it is not an exceptional situation. Then maybe `deleteFuture.complete()` would be better instead of closing a `Watch`

Reply:
@sleshchenko  It appears (also through debugging) that it's clearly not the case: when the POD doesn't exist, the `DeleteWatcher` isn't called at all by the `WatchConnectionManager`  and finally the `WatchConnectionManager` is never closed.

This is precisely the bug I'm fixing here.

This is why I have to explicitly close the `WatchConnectionManager` when the `delete()` call returns `false` because no such POD exists.

Reply:
I understand the fixed bug and it's nice catch =)
And what I understand is (as far as I understand) if you completed future then onComplete will be called and watch will be closed. And completing the future without exception would be more correctly here.

Reply:
well, we would still have to close the watch in case an exception occurs in the `podResource.delete()` call and no `DeleteWatcher` method is called. So why not do the same action in the 2 cases where deletion didn't occur ?